### PR TITLE
Jest testing for use cases and Refactoring

### PR DESCRIPTION
**Pull request by aidanmrli**
Implemented Jest testing for all use cases. The test files can be found in the server/src/__test__ folders. To run the tests, go to the server directory and run npm test.

Additionally, revamped and refactored existing use cases: added catch errors, extracted methods in various files to improve code smells and reduce large chunks of code.

This approach to testing use-cases requires calls to be made to MongoDB - this is unavoidable since our User "entity" is a Mongoose schema. To reduce making costly and slow calls to a real database as the program grows larger, we create a MongoDB memory server, storing test data in memory and separate from our actual user data.

To continue, we would like to develop a test suite for the controllers and for the front-end on the client-side eventually. The use-cases will always be a priority, and we can also think about making better use of Jest Mocking for easier and better designed tests.

### CODE REVIEW

File: server/src/UseCases/TranscriptUseCases/transcriptUploadInteractor.js [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
 const uploadTranscriptInteractor = async (user, file, filename) => {
+    if (user.transcripts === undefined || user.email === undefined) { throw new Error("Not a valid user") }
     let transcriptNames = []
     const existingTranscriptInfo = Object.entries(user.transcripts)
     try {
         existingTranscriptInfo.forEach(info => {
             transcriptNames.push(Object.entries(info[1])[0][0])
             //info[1].forEach(subInfo => {
-                //transcriptNames.push(subInfo.keys()[0])})
+            //transcriptNames.push(subInfo.keys()[0])})
```
Comment (Dhairya-Khara) [inline server/src/UseCases/TranscriptUseCases/transcriptUploadInteractor.js:11]: We can remove these comments (they are unnecessary and adding to code smells)

File: server/src/__tests__/UseCases/GetUseCases/getOneTranscriptIntentsInteractor.test.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+describe('getTranscriptsInteractor test', () => {
+    it('correctly returns "No intents" if there are no transcripts', () => {
+        const mockUser = { email: 'testing@gmail.com', password: 'solid', transcripts: [] };
+        const user = new User(mockUser);
+        const transcriptName = 'transcript3.json'
+        expect(getOneTranscriptIntentsInteractor(user, transcriptName)).toBe("No intents");
+    });
+
+    // it correctly returns a user's transcript's intents
```
Comment (Dhairya-Khara) [inline server/src/__tests__/UseCases/GetUseCases/getOneTranscriptIntentsInteractor.test.js:16]: These comments can also be removed.

File: server/src/__tests__/UseCases/TranscriptUseCases/transcriptUploadInteractor.test.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    });
+
+    it('stops a duplicate transcript from uploading', async () => {
+        const mockUser = { email: 'testing@gmail.com', password: 'solid', transcripts: [] };
+        const user = new User(mockUser);
+        await user.save()
+        const sampleTranscriptJSON = JSON.stringify(sampleTranscript);
+        const sampleTranscriptFilename = 'transcript1.json'
+        await expect(uploadTranscriptInteractor(user, sampleTranscriptJSON, sampleTranscriptFilename)).resolves;
+        // await expect(uploadTranscriptInteractor(user, sampleTranscriptJSON, sampleTranscriptFilename)).rejects;
```
Comment (Dhairya-Khara) [inline server/src/__tests__/UseCases/TranscriptUseCases/transcriptUploadInteractor.test.js:47]: comment!

**Review by Dhairya-Khara — CHANGES_REQUESTED** [review]
Great work guys. The only feedback I have is that in many files, there is commented code. We should remove those to avoid adding to code smells.

**Comment by aidanmrli** [comment]
> Great work guys. The only feedback I have is that in many files, there is commented code. We should remove those to avoid adding to code smells.

Thank you for the feedback, this is a good point that I missed out. Thank you for catching this. The changes have been made.

**Pull request merged**